### PR TITLE
cffi: Support named type + Pass 'seq.go' test.

### PR DESCRIPTION
**Pull request by corona10**
* Pass 'named.go' test.
* Support to generate named type includes slice and array types.
* Add doc.go for bind package.
* Add comments.
* Support __iadd__ for Go slice types.
* Add a 'seq.go' test for the CFFI engine.

Fix: go-python/gopy#121

### CODE REVIEW

**Comment by corona10** [comment]
@alclaude  Ready for review :)

**Comment by corona10** [comment]
@sbinet I will send a new PR which is about [seqs.go](https://github.com/go-python/gopy/blob/master/_examples/seqs/seqs.go) after this PR is merged.

File: bind/doc.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+// Copyright 2017 The go-python Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+/*
```
Comment (sbinet) [inline bind/doc.go]: Standard documentation for packages look like:

```go
// Package bind provides tools to generate bindings to use Go from Python.
package bind
```

File: bind/doc.go [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+bind package generates language bindings that make it possible
+to call Go code and pass objects from Python.
+
+*/
+
```
Comment (sbinet) [inline bind/doc.go]: remove that empty line (otherwise the documentation isn't attached to the package)

File: bind/gencffi_cdef.go [hunk 3]
```diff
@@ -5,18 +5,22 @@
 package bind
 
 import (
+	"fmt"
 	"go/types"
 	"strconv"
 	"strings"
 )
 
+// Generate C definitions of Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: the canonical way to document functions/methods is:

```go
// genCdefStruct generates C definitions for a Go struct.
func (g *cffiGen) genCdefStruct(s Struct) {
```

(ie: `// [method or function name] [verb] [stuff...]`)

File: bind/gencffi_cdef.go [hunk 4]
```diff
@@ -5,18 +5,22 @@
...
 func (g *cffiGen) genCdefStruct(s Struct) {
 	g.wrapper.Printf("// A type definition of the %[1]s.%[2]s for wrapping.\n", s.Package().Name(), s.sym.cgoname)
 	g.wrapper.Printf("typedef void* %s;\n", s.sym.cgoname)
 	g.wrapper.Printf("extern void* cgo_func_%s_new();\n", s.sym.id)
 }
 
+// Generate C definitions of Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: please use the canonical way to document functions/methods.

File: bind/gencffi_cdef.go [hunk 5]
```diff
@@ -5,18 +5,22 @@
...
 func (g *cffiGen) genCdefType(sym *symbol) {
+	g.wrapper.Printf("// C definitions of the %[1]s.%[2]s.\n", g.pkg.pkg.Name(), sym.id)
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: I would rephrase as:
```
"// C definitions for Go type %[1]s.%[2]s.\n"
```

File: bind/gencffi_cdef.go [hunk 6]
```diff
@@ -29,14 +33,64 @@ func (g *cffiGen) genCdefType(sym *symbol) {
...
+				))
+			}
+			g._genCdefFunc(sym, msym)
+		}
+	}
+	g.genCdefTypeTPStr(sym)
+	g.wrapper.Printf("\n")
+}
+
+// Generate cgo_func_XXXX_new() of Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 7]
```diff
@@ -29,14 +33,64 @@ func (g *cffiGen) genCdefType(sym *symbol) {
...
+	case sym.isInterface():
+	default:
+		panic(fmt.Errorf(
+			"gopy: cdef for %s not handled",
+			sym.gofmt(),
+		))
 	}
 }
 
+// Generate a C definition of Go function.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 8]
```diff
@@ -68,6 +122,52 @@ func (g *cffiGen) genCdefFunc(f Func) {
 	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, f.id, paramString)
 }
 
+// Generate a C definition of Go type method.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 9]
```diff
@@ -68,6 +122,52 @@ func (g *cffiGen) genCdefFunc(f Func) {
...
+func (g *cffiGen) _genCdefFunc(sym *symbol, fsym *symbol) {
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: the name of this method isn't too Go-idiomatic (looks more like python :))
for symmetry with `genTypeCdefInit`, I would perhaps call it `genTypeCdefFunc`, or `genTypeCdefMeth`.

we don't like underscores very much :)

File: bind/gencffi_cdef.go [hunk 10]
```diff
@@ -68,6 +122,52 @@ func (g *cffiGen) genCdefFunc(f Func) {
...
+	switch res.Len() {
+	case 0:
+		cdef_ret = "void"
+	case 1:
+		ret := res.At(0)
+		cdef_ret = g.pkg.syms.symtype(ret.Type()).cgoname
+	}
+
+	if isMethod {
+		selfParam := sym.cgoname + " p0"
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: I would just write this as:
```go
params = append(params, sym.cgoname + " p0")
```

File: bind/gencffi_cdef.go [hunk 11]
```diff
@@ -68,6 +122,52 @@ func (g *cffiGen) genCdefFunc(f Func) {
...
+			index++
+		}
+		paramVar := g.pkg.syms.symtype(arg.Type()).cgoname + " " + "p" + strconv.Itoa(index)
+		params = append(params, paramVar)
+	}
+	paramString := strings.Join(params, ", ")
+	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, fsym.id, paramString)
+}
+
+// Generate a C definition of Go Struct method.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 12]
```diff
@@ -99,6 +199,7 @@ func (g *cffiGen) genCdefMethod(f Func) {
 	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, f.id, paramString)
 }
 
+// Generate C definitions of Getter/Setter for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 13]
```diff
@@ -116,6 +217,7 @@ func (g *cffiGen) genCdefStructMemberGetter(s Struct, i int, f types.Object) {
 	}
 }
 
+// Generate C defintion of Setter for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 14]
```diff
@@ -126,6 +228,12 @@ func (g *cffiGen) genCdefStructMemberSetter(s Struct, i int, f types.Object) {
 	g.wrapper.Printf("extern void cgo_func_%[1]s_setter_%[2]d(void* p0, %[3]s p1);\n", s.sym.id, i+1, cdef_value)
 }
 
+// Generate a C definitions of str method for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_cdef.go [hunk 15]
```diff
@@ -126,6 +228,12 @@ func (g *cffiGen) genCdefStructMemberSetter(s Struct, i int, f types.Object) {
...
 func (g *cffiGen) genCdefStructTPStr(s Struct) {
 	g.wrapper.Printf("extern GoString cgo_func_%[1]s_str(void* p0);\n", s.sym.id)
 }
+
+// Generate a C definitions of str method for Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: doc.

File: bind/gencffi_func.go [hunk 16]
```diff
@@ -253,3 +254,76 @@ func (g *cffiGen) genFuncBody(f Func) {
 		panic(fmt.Errorf("gopy: Not yet implemeted for multiple return."))
 	}
 }
+
+func (g *cffiGen) _genFunc(sym *symbol, fsym *symbol) {
```
Comment (sbinet) [inline bind/gencffi_func.go]: `_genFunc` doesn't look too Go-idiomatic, we don't like underscores :)

File: bind/gencffi_func.go [hunk 17]
```diff
@@ -253,3 +254,76 @@ func (g *cffiGen) genFuncBody(f Func) {
...
+		fsym.goname,
+		strings.Join(funcArgs, ", "),
+		fsym.doc,
+	)
+	g.wrapper.Indent()
+	g._genFuncBody(sym, fsym)
+	g.wrapper.Outdent()
+}
+
+func (g *cffiGen) _genFuncBody(sym *symbol, fsym *symbol) {
```
Comment (sbinet) [inline bind/gencffi_func.go]: we don't like underscores :)

File: bind/gencffi_type.go [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+// license that can be found in the LICENSE file.
+
+package bind
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// Generate Go type with Python wrapper class.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+	g.genTypeInit(sym)
+	g.genTypeMethod(sym)
+	g.genTypeTPStr(sym)
+	if sym.isSlice() {
+		g.genTypeIAdd(sym)
+	}
+	g.wrapper.Outdent()
+}
+
+// Generate Type __iadd__.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+func (g *cffiGen) genTypeIAdd(sym *symbol) {
+	g.wrapper.Printf("def __iadd__(self, value):\n")
+	g.wrapper.Indent()
+	switch {
+	case sym.isSlice():
+		g.wrapper.Printf("if not isinstance(value, collections.Sequence):\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: I don't think that's the test we want.

IIRC, python's `__iadd__` tests whether the value is iterable and complains if not:

```
$> python2
Python 2.7.13 (default, Jul 21 2017, 03:24:34) 
[GCC 7.1.1 20170630] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> lst=[]
>>> lst+=[1,2]
>>> lst
[1, 2]
>>> lst+=1
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: 'int' object is not iterable
```

so I would test for `collections.Iterable` instead.
(and you actually do use the iteration protocol later on with `for elt in value:`)

File: bind/gencffi_type.go [hunk 21]
```diff
@@ -0,0 +1,198 @@
...
+		typ := sym.GoType().Underlying().(*types.Slice)
+		esym := g.pkg.syms.symtype(typ.Elem())
+		g.wrapper.Printf("for elt in value:\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(elt)\n", esym.py2c)
+		g.wrapper.Printf("_cffi_helper.lib.cgo_func_%[1]s_append(self.cgopy, pyitem)\n", sym.id)
+		g.wrapper.Outdent()
+	default:
+		panic(fmt.Errorf(
+			"gopy: init for %s not handled",
```
Comment (sbinet) [inline bind/gencffi_type.go]: looks like copy-paste from the `__init__` handling.

File: bind/gencffi_type.go [hunk 22]
```diff
@@ -0,0 +1,198 @@
...
+			sym.gofmt(),
+		))
+	}
+	g.wrapper.Printf("return self\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __init__.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 23]
```diff
@@ -0,0 +1,198 @@
...
+		g.wrapper.Printf("if len(args) == 1:\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("self.cgopy = _cffi_helper.cffi_%[1]s(args[0])\n",
+			bsym.py2c,
+		)
+		g.wrapper.Outdent()
+	case sym.isArray():
+		g.wrapper.Printf("if args:\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("if not isinstance(args[0], collections.Sequence):\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: here, `collections.Sequence` is ok. (we need `len(x)`)

File: bind/gencffi_type.go [hunk 24]
```diff
@@ -0,0 +1,198 @@
...
+		g.wrapper.Printf("for idx, elt in enumerate(args[0]):\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("pyitem = _cffi_helper.cffi_%[1]s(elt)\n", esym.py2c)
+		g.wrapper.Printf("_cffi_helper.lib.cgo_func_%[1]s_ass_item(self.cgopy, idx, pyitem)\n", sym.id)
+		g.wrapper.Outdent()
+		g.wrapper.Outdent()
+	case sym.isSlice():
+		g.wrapper.Printf("if args:\n")
+		g.wrapper.Indent()
+		g.wrapper.Printf("if not isinstance(args[0], collections.Sequence):\n")
```
Comment (sbinet) [inline bind/gencffi_type.go]: but here, I would use `collections.Iterable` instead.

File: bind/gencffi_type.go [hunk 25]
```diff
@@ -0,0 +1,198 @@
...
+		panic(fmt.Errorf(
+			"gopy: init for %s not handled",
+			sym.gofmt(),
+		))
+	}
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type converter between C and Python.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 26]
```diff
@@ -0,0 +1,198 @@
...
+	g.wrapper.Printf("\n")
+	g.wrapper.Printf("@staticmethod\n")
+	g.wrapper.Printf("def cffi_cgopy_cnv_c2py_%[1]s(c):\n", sym.id)
+	g.wrapper.Indent()
+	g.wrapper.Printf("raise NotImplementedError\n")
+	g.wrapper.Outdent()
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type methods.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

File: bind/gencffi_type.go [hunk 27]
```diff
@@ -0,0 +1,198 @@
...
+					m.Name()+" || "+m.FullName(),
+				))
+			}
+			g._genFunc(sym, msym)
+		}
+	}
+	g.wrapper.Printf("\n")
+}
+
+// Generate Type __str__ method.
```
Comment (sbinet) [inline bind/gencffi_type.go]: doc.

**Comment by corona10** [comment]
@sbinet Updated!!

File: bind/doc.go [hunk 28]
```diff
@@ -0,0 +1,7 @@
+// Copyright 2017 The go-python Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package bind package generates language bindings that make it possible
```
Comment (sbinet) [inline bind/doc.go]: typo: `Package bind package`

File: bind/gencffi_cdef.go [hunk 29]
```diff
@@ -5,18 +5,22 @@
 package bind
 
 import (
+	"fmt"
 	"go/types"
 	"strconv"
 	"strings"
 )
 
+// genCdefStruct generate C definitions of Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/generate/generates/`
`s/of Go/for a Go/`

File: bind/gencffi_cdef.go [hunk 30]
```diff
@@ -5,18 +5,22 @@
...
 func (g *cffiGen) genCdefStruct(s Struct) {
 	g.wrapper.Printf("// A type definition of the %[1]s.%[2]s for wrapping.\n", s.Package().Name(), s.sym.cgoname)
 	g.wrapper.Printf("typedef void* %s;\n", s.sym.cgoname)
 	g.wrapper.Printf("extern void* cgo_func_%s_new();\n", s.sym.id)
 }
 
+// genCdefType generates C definitions of Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/of Go/for a Go/`

File: bind/gencffi_cdef.go [hunk 31]
```diff
@@ -29,14 +33,64 @@ func (g *cffiGen) genCdefType(sym *symbol) {
...
+				))
+			}
+			g.genTypeCdefFunc(sym, msym)
+		}
+	}
+	g.genCdefTypeTPStr(sym)
+	g.wrapper.Printf("\n")
+}
+
+// genTypeCdefInit generate cgo_func_XXXX_new() of Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/generate/generates/`
`s/of Go/for a Go/`

File: bind/gencffi_cdef.go [hunk 32]
```diff
@@ -29,14 +33,64 @@ func (g *cffiGen) genCdefType(sym *symbol) {
...
+	case sym.isInterface():
+	default:
+		panic(fmt.Errorf(
+			"gopy: cdef for %s not handled",
+			sym.gofmt(),
+		))
 	}
 }
 
+// genCdefFunc generates a C definition of Go function.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/of Go/for a Go/`

File: bind/gencffi_cdef.go [hunk 33]
```diff
@@ -68,6 +122,51 @@ func (g *cffiGen) genCdefFunc(f Func) {
 	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, f.id, paramString)
 }
 
+// genTypeCdefFunc generate a C definition of Go type method.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/generate/generates/`
`s/of Go type method/for a Go type's method/`

File: bind/gencffi_cdef.go [hunk 34]
```diff
@@ -68,6 +122,51 @@ func (g *cffiGen) genCdefFunc(f Func) {
...
+			index++
+		}
+		paramVar := g.pkg.syms.symtype(arg.Type()).cgoname + " " + "p" + strconv.Itoa(index)
+		params = append(params, paramVar)
+	}
+	paramString := strings.Join(params, ", ")
+	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, fsym.id, paramString)
+}
+
+// genCdefMethod generates a C definition of Go Struct method.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/of Go Struct method/for a Go struct's method/`

File: bind/gencffi_cdef.go [hunk 35]
```diff
@@ -99,6 +198,8 @@ func (g *cffiGen) genCdefMethod(f Func) {
 	g.wrapper.Printf("extern %[1]s cgo_func_%[2]s(%[3]s);\n", cdef_ret, f.id, paramString)
 }
 
+// genCdefStructMemberGetter generates
+// C definitions of Getter/Setter for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/for Go struct/for a Go struct/`
also, please everything on one line (ie: no new line after `generates`.)

File: bind/gencffi_cdef.go [hunk 36]
```diff
@@ -116,6 +217,8 @@ func (g *cffiGen) genCdefStructMemberGetter(s Struct, i int, f types.Object) {
 	}
 }
 
+// genCdefStructMemberSetter generates
+// C defintion of Setter for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/for Go/for a Go/`.
`s/defintion/definition/`
also: please everything on one line.

File: bind/gencffi_cdef.go [hunk 37]
```diff
@@ -126,6 +229,14 @@ func (g *cffiGen) genCdefStructMemberSetter(s Struct, i int, f types.Object) {
 	g.wrapper.Printf("extern void cgo_func_%[1]s_setter_%[2]d(void* p0, %[3]s p1);\n", s.sym.id, i+1, cdef_value)
 }
 
+// genCdefStructTPStr generates
+// C definitions of str method for Go struct.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/for Go/for a Go/`
also: please everything on one line.

File: bind/gencffi_cdef.go [hunk 38]
```diff
@@ -126,6 +229,14 @@ func (g *cffiGen) genCdefStructMemberSetter(s Struct, i int, f types.Object) {
...
 func (g *cffiGen) genCdefStructTPStr(s Struct) {
 	g.wrapper.Printf("extern GoString cgo_func_%[1]s_str(void* p0);\n", s.sym.id)
 }
+
+// genCdefTypeTPStr generates
+// C definitions of str method for Go type.
```
Comment (sbinet) [inline bind/gencffi_cdef.go]: `s/for Go/for a Go/`
also: please everything on one line.

File: bind/gencffi_type.go [hunk 39]
```diff
@@ -0,0 +1,198 @@
...
+// license that can be found in the LICENSE file.
+
+package bind
+
+import (
+	"fmt"
+	"go/types"
+)
+
+// genType generates Go type with Python wrapper class.
```
Comment (sbinet) [inline bind/gencffi_type.go]: that doesn't read very well.
please replace with something along the lines of:
```go
// genType generates a Python class wrapping the given Go type.
```

**Comment by corona10** [comment]
@sbinet 
Thank you for comment about all sentences.
The update was a little bit late due to unexpected event.

**Comment by corona10** [comment]
Updated
